### ISSUE TITLE
Unable to use TLS private key in secure storage

### ISSUE THREAD

**Issue by snej** [warmcat/libwebsockets#1652]
**tl;dr:** I request an API for LWS to support an external TLS private key that's accessed through client-supplied crypto callbacks.

The project I work on needs to keep TLS private keys in the OS trusted storage module, i.e. the CPU "secure enclave". This requires using OS calls to do the encryption/signing. The key data cannot leave secure storage, so it's not possible to pass it to libwebsockets directly nor write it to a file. (On iOS and macOS this secure enclave is accessed via the Keychain API; I know there are equivalents on Android, other Linuxes and Windows. And a lot of newer embedded SoCs have a similar facility, as IoT security becomes a bigger deal.)

mbedTLS [supports this](https://tls.mbed.org/kb/how-to/ssl_async) — you can either create an "RSA alt" context that behaves like a regular key-pair but delegates the crypto to custom callbacks; or you can register crypto callbacks at the SSL context level via `mbedtls_ssl_conf_async_private_cb`.

(I'm less familiar with OpenSSL, but it has an "engine" API that looks like it does the same thing, basically "subclassing" their key type to add custom operations. I'm not sure if there's a higher level API at the SSL-context level.)

I've gotten the mbedTLS external-key support working on its own, but can't find any way to use it with libwebsockets. The TLS context isn't exposed through the API, and there's another level of indirection since LWS goes through an OpenSSL compatibility shim. I've spent a few days looking for a clean way to hack LWS to add this, but can't find one (partly because I get lost in the twisty maze of little passages in the 3rd party OpenSSL adapter code.)

**Comment by snej** [comment]
I wonder if I can use `LWS_CALLBACK_OPENSSL_LOAD_EXTRA_SERVER_VERIFY_CERTS` to do this, since it gives me a pointer to the SSL context. (You know how you sometimes find the answer to a question right after you post it? Might be happening here.) It would still be a bit of a hack, especially since I'd have to grope the mbedTLS context out of the fake OpenSSL one.

**Comment by lws-team** [comment]
The tls context is deliberately opaque in lws so you can swap between mbedtls and openssl backends cleanly.  There's even a pile of gencrypto apis you can use that acts as a single api that works the same on top of either tls kibrary's completely different apis.

Like ENGINE this is a tls-lib level extension that is completely different between the libs, because each went their own sweet way.  So this code's natural place to live is down ./lib/tls/mbedtls and to the extent any interface is needed to other bits, it needs to go through a generic, tls-lib agnostic api.

Can you sketch out what the mbedtls implementation looks like with mbedtls apis + pseudocode, and the pointsbof contact it has outside its box?  Eg, you have to send it a privkey index along with the sign request?  It also handles pubketys / verify?  RSA?  EC?

**Comment by snej** [comment]
We already use mbedTLS directly, and I do not want to rewrite everything to a different compatibility API.

I understand the desire to provide abstractions, but the client app configures LWS to use one TLS engine and therefore knows exactly which one is being used. And if OpenSSL is used, the context is not opaque — the callback docs say it's "the server's OpenSSL SSL_CTX*".

The missing piece is APIs to allow the app to work with that context if it's using mbedTLS. In my clone I've added some functions to lws.h:

```c
#if defined(LWS_WITH_MBEDTLS)
/* SSL/TLS context access, for use with LWS_CALLBACK_OPENSSL_LOAD_EXTRA_CLIENT_VERIFY_CERTS and
   LWS_CALLBACK_OPENSSL_LOAD_EXTRA_SERVER_VERIFY_CERTS. In these callbacks, the 'user' pointer
   may be cast to 'lws_ssl_ctx*' and passed to these functions. */

typedef struct lws_ssl_ctx lws_ssl_ctx;

mbedtls_x509_crt *
lws_ssl_ctx_get_mbedtls_x509_crt(lws_ssl_ctx *ssl_ctx);

// Copies the cert, so you may free it afterwards
int lws_ssl_ctx_set_mbedtls_x509_crt(lws_ssl_ctx *ssl_ctx, mbedtls_x509_crt *crt);

mbedtls_pk_context *
lws_ssl_ctx_get_mbedtls_key(lws_ssl_ctx *ssl_ctx);

int lws_ssl_ctx_set_mbedtls_key(lws_ssl_ctx *ssl_ctx, mbedtls_pk_context *key);

#endif
```

**Comment by lws-team** [comment]
Nobody is suggesting 'rewriting everything under a different compatibility api'.

The point is the parts that come out of the box must expose themselves to the user equally no matter the tls backend... that means not passing the SSL_CTX around.  For the crypto apis they all come out of the box, they all have different key layouts and different context api semantics, and the only way is to make wrappers.

For this most of it stays in the box.   The mbedtls bits stay in ./lib/tls/mbedtls as I said before.  But I'm done adding single-purpose hack apis around tls as you're suggesting.  I have spent months doing the work to make lws tls backend agnostic to get away from that.

Why don't we accept we each have a different perspective on this and discuss what's needed to do this in a way that doesn't expose these entrails.

**Comment by lws-team** [comment]
... let me expand a bit on my meaning... if I can't convince you then maybe someone else interested can find it by google.

It looks like the service provider like this keychain api should become a thing with its own ops struct that tries to normalize operations from different kinds of providers.... eg lws_crypto_service_t.  That's useful in itself if you expect to deploy on more than one platform.  You can pass a pointer to the lws_crypto_service_t either per context or per vhost when you create it.

Then there would be code in ./lib/tls/mbedtls and ./lib/tls/openssl to interface to the function pointers in lws_crypto_service_t to get stuff done, if it had been created with it set to non-NULL.  Since you have some code already you know more than me but generally the event where lws wants to perform these ops I should think already exists.  It seems it's just a matter of getting the normal operations done via this service provider.

**Comment by snej** [comment]
Let's see, you could implement this as follows:

1. Extend `lws_context_creation_info` to allow the key-pair to be provided as an object. I see you've added structs for RSA keys and EC keys, but they're incompatible; I'm thinking more of mbedTLS's `mbedtls_pk_context` which abstracts all keys.
2. Add a function to create a key from a set of callbacks; this is what `mbedtls_pk_setup_rsa_alt` does. The parameters are the key size, callback functions to sign and decrypt, and an opaque context to be passed to the callbacks.

Then I'd call the key-creation function, passing in callbacks that call the Keychain APIs to do the crypto, plus a handle to the private-key object in the Keychain. The resulting LWS key pointer would go into the `lws_context_creation_info`.

Here's the prototype of the mbedTLS function:
```c
int mbedtls_pk_setup_rsa_alt( mbedtls_pk_context *ctx, void * key,
                         mbedtls_pk_rsa_alt_decrypt_func decrypt_func,
                         mbedtls_pk_rsa_alt_sign_func sign_func,
                         mbedtls_pk_rsa_alt_key_len_func key_len_func );
```
and the callbacks are defined as
```
typedef int (*mbedtls_pk_rsa_alt_decrypt_func)( void *ctx, int mode, size_t *olen,
                    const unsigned char *input, unsigned char *output,
                    size_t output_max_len );
typedef int (*mbedtls_pk_rsa_alt_sign_func)( void *ctx,
                    int (*f_rng)(void *, unsigned char *, size_t), void *p_rng,
                    int mode, mbedtls_md_type_t md_alg, unsigned int hashlen,
                    const unsigned char *hash, unsigned char *sig );
typedef size_t (*mbedtls_pk_rsa_alt_key_len_func)( void *ctx );
```
(Note: I think it's silly to use a callback for the key length, since it's a constant. Simpler just to pass it in to the creation function and store it in the internal struct.)

**Comment by snej** [comment]
Step 2 I could do easily, but I'm unsure about step 1 since it involves some API refactoring to create a common abstract key type instead of using separate structs for RSA and EC. Or I guess instead you could extend the creation_info with an RSA pointer, an EC pointer and an external-key pointer... 😝

In the meantime, I'm forking LWS slightly to add the functions I showed yesterday. I've successfully gotten TLS connections working using secure keys from the Keychain, so I'm happy for now.

**Comment by lws-team** [comment]
> added structs for RSA keys and EC keys, but they're incompatible; 

The gencrypto stuff unifies mbedtls and openssl rsa and ec ops... they devolve to whatever lib you're using, but the apis are the same either way.

https://libwebsockets.org/git/libwebsockets/tree/include/libwebsockets/lws-gencrypto.h
https://libwebsockets.org/git/libwebsockets/tree/include/libwebsockets/lws-genec.h
https://libwebsockets.org/git/libwebsockets/tree/include/libwebsockets/lws-genrsa.h
... and for hashes and aes.

Well if you want to fork it and don't really care, no worries.

**Comment by snej** [comment]
>The gencrypto stuff unifies mbedtls and openssl rsa and ec ops

What I meant is that your API doesn't unify RSA and EC keys as a single abstract "key-pair" type. They're separate incompatible structs. Without that, what I'm proposing won't work, because it effectively creates a 3rd type of underlying key ("external") hiding behind the abstraction. mbedTLS is able to use external keys because it has an abstract `mbedtls_pk_context` type.

Why did you close the issue? It's not resolved, and it sounded like we were converging on a solution.

**Comment by lws-team** [comment]
Honestly what you pasted seems the same as what you suggested before.  Please find a moment to reread and consider this I posted earlier:

> It looks like the service provider like this keychain api should become a thing with its own ops struct that tries to normalize operations from different kinds of providers.... eg lws_crypto_service_t. That's useful in itself if you expect to deploy on more than one platform. You can pass a pointer to the lws_crypto_service_t either per context or per vhost when you create it.  Then there would be code in ./lib/tls/mbedtls and ./lib/tls/openssl to interface to the function pointers in lws_crypto_service_t to get stuff done, if it had been created with it set to non-NULL. Since you have some code already you know more than me but generally the event where lws wants to perform these ops I should think already exists. It seems it's just a matter of getting the normal operations done via this service provider.

It means there should be no exports with "mbedtls" in the name... it's the wrong direction.  Instead there should be a standardized way to represent the **crypto service** via an external ops struct passed in at context or vhost creation time, which can be used the same by mbedtls or openssl.  That way when you meet a second, incompatible crypto service, ideally the amount of work that needs doing is limited to writing the ops struct shim for the new service (plus adapting it to any wildly different semantics).  By then mbedtls and openssl already know how to use the ops to get stuff done and should only need minimal work.

**Comment by snej** [comment]
> Honestly what you pasted seems the same as what you suggested before

It's not. I think you may have been misled by the snippets of the mbedTLS headers that I posted; that is _not_ what I'm proposing to add, rather the mechanism used to implement it for the mbedTLS backend.

A quick sketch of what I'd propose adding to LWS is:

```c
// Opaque key reference. Existing RSA and ECC key types would also be abstracted to this.
typedef struct lws_key lws_key;

struct lws_context_creation_info {
    ...
    lws_key *server_ssl_private_key;    // server private key as a direct reference
    lws_key *client_ssl_private_key;    // client private key as a direct reference
    ...
};

// Creates a reference to an external key that's accessed via callbacks.
// This allows for the use of keys managed by OS security APIs, CPU secure enclaves,
// or external crypto dongles like YubiKey.
lws_key* lws_create_external_private_key(
    void * external_key_ref,
    unsigned key_length,
    lws_decrypt_func decrypt_func,
    lws_sign_func sign_func);
```

I'm not showing the decript/sign callbacks; they'd be basically identical to the mbedTLS ones I gave above.
